src/content/docs/en/reference/loader-reference.mdx

```diff
@@ -3,6 +3,324 @@ title: Astro Loader API
+
+```ts
+const countries = defineCollection({
+  loader: async () => {
+    const response = await fetch("https://restcountries.com/v3.1/all");
+    const data = await response.json();
+    // Must return an array of entries with an id property
+    // or an object with IDs as keys and entries as values
+    return data.map((country) => ({
+      id: country.cca3,
+      ...country,
+    }));
+  },
+});
+```
+
+### Object loaders
+
+If a loader needs more control over the loading process it can use the object loader API. The loader is defined as an object with a `load()` method that is called at build time to fetch data and update the data store. It allows entries to be updated incrementally, or for the store to be cleared only when necessary. It can also define a schema for the entries, which can be used to validate the data and generate static types. 
+
+The recommended pattern is to define a function that accepts configuration options and returns the loader object, in the same way that you would normally define an Astro integration or Vite plugin. 
+
+
+```ts title=loader.ts
+
+import type { Loader, LoaderContext } from 'astro/loaders';
+
+// Define any options that the loader needs
+export function myLoader(options: { url: string }): Loader {
+  // Configure the loader 
+  const feedUrl = new URL(options.url);
+  // Return a loader object
+  return {
+    name: "my-loader";
+    // Called when updating the collection.
+    load: async (context: LoaderContext): Promise<void> => {
+      // Load data and update the store
+    }
+    // Optionally, define the schema of an entry. 
+    // It will be overridden by user-defined schema. 
+    schema: async () => z.object({
+      // ...
+    })
+  };
+}
+```
+
+## Loader API
+
+The API for [inline loaders](#inline-loaders) is very simple, and is shown above. This section shows the API for defining an object loader.
+
+### The `Loader` object
+
+The loader object has the following properties:
+
+#### `name`
+**Type**: `string`
+
+A unique name for the loader, used in logs and [for conditional loading](#integration-support).
+
+#### `load`
+**Type**: `(context: LoaderContext) => Promise<void>`
+
+An async function that is called at build time to load data and update the store. See [`LoaderContext`](#loadercontext) for more information.
+
+#### `schema`
+**Type**: `ZodSchema | Promise<ZodSchema> | (() => ZodSchema | Promise<ZodSchema>)`
+
+An optional [Zod schema](/en/guides/defining-collections/#defining-datatypes-with-zod) that defines the shape of the entries. It is used to both validate the data, and also to generate TypeScript types for the collection. If a function is provided, it will be called at build time before `load()` to generate the schema. You can use this to dynamically generate the schema based on the configuration options or by introspecting an API. 
```

Comment:
> It is used to both validate the data, and also to generate TypeScript types for the collection.
> [...]
> or by introspecting an API. 

I think it is a bit weird to have the API introspection example when stating that this is used to generate the types because an introspected schema won't be typed.

Reply:
It would. You can e.g. do an introspection query to an API endpoint, and then generate the schema from that. This will then in turn generate the TypeScript types from that schema.